### PR TITLE
journald: remove span field prefixes; add separate fields for span data

### PR DESCRIPTION
**Pull request by wiktorsikora**
## Motivation
Currently, `tracing-journald` prefixes event fields with the number
of parent spans, if present, in order to namespace field values. This
turned out to be unnecessary as journald preserves values for duplicated
field names.

## Solution
This branch removes event field prefixes and emits parent span names
and their field/value pairs as additional key/value pairs.

### CODE REVIEW

File: tracing-journald/src/lib.rs [hunk 1]
```diff
@@ -277,16 +288,19 @@ struct SpanFields(Vec<u8>);
...
 impl SpanVisitor<'_> {
     fn put_span_prefix(&mut self) {
-        write!(self.buf, "S{}", self.depth).unwrap();
-        if let Some(prefix) = self.prefix {
-            self.buf.extend_from_slice(prefix.as_bytes());
+        if let Some(span_prefix) = self.span_field_prefix {
+            write!(self.buf, "{}{}", span_prefix, self.depth).unwrap();
+            if let Some(prefix) = self.field_prefix {
+                self.buf.extend_from_slice(prefix.as_bytes());
+            }
```
Comment (hawkw) [inline tracing-journald/src/lib.rs:301]: it looks like if `span_field_prefix` is `None`, we will also skip emitting the field prefix. I'm not sure if that is correct --- this would mean there is no way to disable the span field prefix without also disabling the field prefix. i _think_ these should be configured separately.

**Review by hawkw — COMMENTED** [review]
thanks for working on this! i noticed a few things we may want to address before this is ready to merge --- in particular, i think the relationship between the span field prefix and field prefix configurations seems a bit surprising.

also, it would be nice to hear from @Ralith before merging this.

File: tracing-journald/src/lib.rs [hunk 2]
```diff
@@ -277,16 +288,19 @@ struct SpanFields(Vec<u8>);
...
+    span_field_prefix: Option<&'a str>,
 }
 
 impl SpanVisitor<'_> {
     fn put_span_prefix(&mut self) {
-        write!(self.buf, "S{}", self.depth).unwrap();
-        if let Some(prefix) = self.prefix {
-            self.buf.extend_from_slice(prefix.as_bytes());
+        if let Some(span_prefix) = self.span_field_prefix {
+            write!(self.buf, "{}{}", span_prefix, self.depth).unwrap();
```
[the remark this answers is not in the document]
Comment (Ralith) [reply tracing-journald/src/lib.rs:298]: I think this is a special case of requiring that `span_prefix` satisfies the requirements of journald, and documenting the conditions under which we prefix the depth. My inclination is to just rely on documentation here.

File: tracing-journald/src/lib.rs [hunk 3]
```diff
@@ -129,6 +131,13 @@ impl Subscriber {
         self
     }
 
+    /// Sets the prefix to apply to names of user-defined span fields.
+    /// Defaults to `Some("S")`.
+    pub fn with_span_field_prefix(mut self, x: Option<String>) -> Self {
+        self.span_field_prefix = x;
```
[the remark this answers is not in the document]
Comment (Ralith) [reply tracing-journald/src/lib.rs:137]: I think we should just document that that requirement, and ideally link an appropriate reference.

File: tracing-journald/src/lib.rs [hunk 4]
```diff
@@ -129,6 +131,13 @@ impl Subscriber {
         self
     }
 
+    /// Sets the prefix to apply to names of user-defined span fields.
+    /// Defaults to `Some("S")`.
```
[the remark this answers is not in the document]
Comment (Ralith) [reply tracing-journald/src/lib.rs:135]: This should be clearer about what happens to the `<DEPTH>_` part of the prefix.

**Comment by Ralith** [comment]
I didn't actually realize journald preserved values for duplicate field names. Maybe we should drop the span prefixes entirely? This would be breaking, but probably gives more intuitive/natural results in the common case.

File: tracing-journald/src/lib.rs (hunk 1 above)
Comment (wiktorsikora) [reply tracing-journald/src/lib.rs]: Sure, that's my oversight.

File: tracing-journald/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wiktorsikora) [reply tracing-journald/src/lib.rs:145]: documented

**Comment by Ralith** [comment]
> I agree it would be less confusing to disable prefixing of the user span fields by default.

What about removing support for prefixing spans entirely? It seems superfluous to me in light of journald's behavior as tested here. I'm not sure anyone actually wants the current behavior, even as an option. I'd also like to get @hawkw's input on the compatibility break, too; if we go this way, it should probably be 0.2 so it doesn't catch anyone relying on the current behavior by surprise.

**Comment by wiktorsikora** [comment]
I removed span fields prefixing. I also added `SPAN_NAME` field emitted for events recorded inside spans. 
As @Ralith pointed out these are breaking changes.

File: tracing-journald/src/lib.rs [hunk 5]
```diff
@@ -345,8 +337,8 @@ impl Visit for EventVisitor<'_> {
     }
 }
 
-fn put_metadata(buf: &mut Vec<u8>, meta: &Metadata, span: Option<usize>) {
-    if span.is_none() {
+fn put_metadata(buf: &mut Vec<u8>, meta: &Metadata, span: bool) {
+    if !span {
```
Comment (Ralith) [inline tracing-journald/src/lib.rs:341]: I think it would make more sense to move the `PRIORITY` field into a separate call (or just hardcode it directly in `on_event`).

File: tracing-journald/tests/journal.rs [hunk 6]
```diff
@@ -32,15 +36,34 @@ fn with_journald(f: impl FnOnce()) {
 #[serde(untagged)]
 enum Field {
     Text(String),
+    Array(Vec<String>),
     Binary(Vec<u8>),
 }
 
+impl Field {
+    pub fn as_array(&self) -> Option<&Vec<String>> {
```
Comment (Ralith) [inline tracing-journald/tests/journal.rs:44]: Nit: `Option<&[String]>`

File: tracing-journald/src/lib.rs [hunk 7]
```diff
@@ -359,20 +351,11 @@ fn put_metadata(buf: &mut Vec<u8>, meta: &Metadata, span: Option<usize>) {
...
-        }
         put_field_wellformed(buf, "CODE_FILE", file.as_bytes());
     }
     if let Some(line) = meta.line() {
-        if let Some(n) = span {
-            write!(buf, "S{}_", n).unwrap();
-        }
         // Text format is safe as a line number can't possibly contain anything funny
         writeln!(buf, "CODE_LINE={}", line).unwrap();
     }
```
Comment (Ralith) [inline tracing-journald/src/lib.rs:361]: `CODE_FILE` and `CODE_LINE` are standard fields that may be interpreted by other off-the-shelf tools. How will they behave when there are multiple values for them? If they generally select the a particular value (perhaps the first or last instance?) perhaps we should make sure that's the event's value. If that's not feasible, should we prefix them with `SPAN_` to avoid confusion?

File: tracing-journald/tests/journal.rs [hunk 8]
```diff
@@ -16,9 +16,13 @@ fn journalctl_version() -> std::io::Result<String> {
...
 
 fn with_journald(f: impl FnOnce()) {
+    with_journald_subscriber(Subscriber::new().unwrap().with_field_prefix(None), f)
+}
+
+fn with_journald_subscriber(subscriber: Subscriber, f: impl FnOnce()) {
     match journalctl_version() {
         Ok(_) => {
-            let sub = Registry::default().with(Subscriber::new().unwrap().with_field_prefix(None));
+            let sub = Registry::default().with(subscriber);
```
Comment (Ralith) [inline tracing-journald/tests/journal.rs:25]: Are these changes still needed?

**Review by Ralith — CHANGES_REQUESTED** [review]

File: tracing-journald/src/lib.rs (hunk 7 above)
Comment (wiktorsikora) [reply tracing-journald/src/lib.rs]: Right, it seems more intuitive to record these fields with span prefix. Fixed

File: tracing-journald/tests/journal.rs (hunk 8 above)
Comment (wiktorsikora) [reply tracing-journald/tests/journal.rs:25]: I believe so. I'm using this in `simple_metadata`  test. It may also be useful in future tests

**Review by Ralith — APPROVED** [review]

**Review by Ralith — APPROVED** [review]

File: tracing-journald/tests/journal.rs [hunk 9]
```diff
@@ -40,6 +40,23 @@ enum Field {
...
+            Field::Text(_) => None,
+            Field::Binary(_) => None,
+            Field::Array(v) => Some(v),
+        }
+    }
+    pub fn as_text(&self) -> Option<&str> {
+        match self {
+            Field::Text(v) => Some(v.as_str()),
+            Field::Binary(_) => None,
+            Field::Array(v) => None,
```
Comment (hawkw) [inline tracing-journald/tests/journal.rs]: nit:
```suggestion
            Field::Array(_) => None,
```

**Comment by wiktorsikora** [comment]
@hawkw is there anything else I could do on this? No pressure, I just want to know :)

**Review by hawkw — APPROVED** [review]
i think this looks good to me, thanks!

**Comment by psztoch** [comment]
When will these changes go to the master?
The proposed changes are VERY desirable.

**Review by davidbarsky — APPROVED** [review]
